src/arch/x86/x86_64/src/bzimage.rs

```diff
@@ -0,0 +1,352 @@
+        bp.e820_entries = 5;
+    }
+
+    /// Run the payload. This might not return.
+    pub fn run(&self, w: &mut print::WriteTo) {
+        // Jump to the payload.
+        // See: linux/Documentation/arm/Booting
+        unsafe {
+            let f = transmute::<usize, EntryPoint>(self.entry);
+            write!(w, "on to {:#x} ", self.entry).unwrap();
+            f(1, DTB_ADDR);
+        }
+        // TODO: error when payload returns.
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn setup_e820_tables_correctly_sets_table() {
+        let test_image = BzImage { low_mem_size: 0xFFFFFF, high_mem_start: 0xFFFF, high_mem_size: 0, rom_base: 0xFF_000_000, rom_size: 0xAAAA, load: 0xAAAA, entry: 0xAAAA };
+        let mut test_bp = BootParams::default();
+        let test_hdr = SetupHeader {
+            setup_sects: 26,
+            root_flags: 1,
+            syssize: 40892,
+            ram_size: 0,
+            vid_mode: 65535,
+            root_dev: 0,
+            boot_flag: 43605,
+            jump: 27371,
+            header: 1400005704,
+            version: 527,
+            realmode_swtch: 0,
+            start_sys: 4096,
+            kernel_version: 12736,
+            type_of_loader: 0,
+            loadflags: 1,
+            setup_move_size: 32768,
+            code32_start: 1048576,
+            ramdisk_image: 0,
+            ramdisk_size: 0,
+            bootsect_kludge: 0,
+            heap_end_ptr: 19056,
+            ext_loader_ver: 0,
+            ext_loader_type: 0,
+            cmd_line_ptr: 0,
+            initrd_addr_max: 2147483647,
+            kernel_alignment: 2097152,
+            relocatable_kernel: 0,
+            min_alignment: 21,
+            xloadflags: 97,
+            cmdline_size: 2047,
+            hardware_subarch: 0,
+            hardware_subarch_data: 0,
+            payload_offset: 740,
+            payload_length: 633540,
+            setup_data: 0,
+            pref_address: 16777216,
+            init_size: 11591680,
+            handover_offset: 0,
+            kernel_info_offset: 652280,
+        };
+        test_bp.hdr = test_hdr;
+        test_image.setup_e820_tables(&mut test_bp);
+
+        //Set up E820Entries for test
+        let entry_first_page = E820Entry { addr: 0, size: PAGE_SIZE as u64, r#type: E820::RESERVED as u32 };
+        let entry_low = E820Entry { addr: PAGE_SIZE as u64, size: (LOW_MEM_64K - PAGE_SIZE) as u64, r#type: E820::RAM as u32 };
+        let entry_reserved = E820Entry { addr: LOW_MEM_64K as u64, size: (LOW_MEM_1M - LOW_MEM_64K) as u64, r#type: E820::RESERVED as u32 };
+        let entry_low_main = E820Entry { addr: LOW_MEM_1M as u64, size: (test_image.low_mem_size - LOW_MEM_1M as u64), r#type: E820::RAM as u32 };
+        let entry_main = E820Entry { addr: test_image.high_mem_start as u64, size: test_image.high_mem_size as u64, r#type: E820::RAM as u32 };
+
+        assert_eq!(test_bp.e820_table[0], entry_first_page);
```

Comment:
Can you do `assert_eq!` between the two structs?

Reply:
Not sure I understand the question. An E820Entry only has primitives so I derived PartialEq and that should match all of them.

Reply:
That looks good! I wonder if there's any way to have a single `assert_eq!` which compares the whole slices against each other. Otherwise, ti's fine.